File: labs/lab03-migration-workflow/python/agent.py
"""Migration Workflow Agent - Core Agent Implementation."""
from state import MigrationState, Phase, MigrationStep
from prompts import ANALYSIS_PROMPT, PLANNING_PROMPT, MIGRATION_PROMPT, VERIFICATION_PROMPT
from typing import Dict, List
import json


class MigrationAgent:
    """Agent that performs multi-step code migration."""

    def __init__(self, llm_client):
        self.llm = llm_client

    def run(self, state: MigrationState) -> MigrationState:
        """Run the migration agent through all phases."""
        while state.phase != Phase.COMPLETE:
            state = self._step(state)
            if state.errors:
                break
        return state

    def _step(self, state: MigrationState) -> MigrationState:
        """Execute one phase of the migration."""
        if state.phase == Phase.ANALYSIS:
            return self._analyze(state)
        elif state.phase == Phase.PLANNING:
            return self._plan(state)
        elif state.phase == Phase.EXECUTION:
            return self._execute(state)
        elif state.phase == Phase.VERIFICATION:
            return self._verify(state)
        return state

    def _analyze(self, state: MigrationState) -> MigrationState:
        """Phase 1: Analyze source code."""
        all_analysis = {}

        for filename, code in state.source_files.items():
            prompt = ANALYSIS_PROMPT.format(
                source=state.source_framework,
                target=state.target_framework,
                language=self._detect_language(filename),
                code=code
            )

            response = self.llm.chat([
                {"role": "user", "content": prompt}
            ])

            try:
                all_analysis[filename] = self._parse_json(response)
            except Exception as e:
                state.errors.append(f"Analysis failed for {filename}: {e}")
                return state

        state.analysis = all_analysis
        state.phase = Phase.PLANNING
        return state

    def _plan(self, state: MigrationState) -> MigrationState:
        """Phase 2: Create migration plan."""
        prompt = PLANNING_PROMPT.format(
            analysis=json.dumps(state.analysis, indent=2),
            source=state.source_framework,
            target=state.target_framework
        )

        response = self.llm.chat([
            {"role": "user", "content": prompt}
        ])

        try:
            plan_data = self._parse_json(response)
            state.plan = [
                MigrationStep(
                    id=step["id"],
                    description=step["description"],
                    input_files=step.get("input_files", [])
                )
                for step in plan_data.get("steps", [])
            ]
        except Exception as e:
            state.errors.append(f"Planning failed: {e}")
            return state

        state.phase = Phase.EXECUTION
        return state

    def _execute(self, state: MigrationState) -> MigrationState:
        """Phase 3: Execute migration steps."""
        while state.current_step < len(state.plan):
            step = state.plan[state.current_step]
            step.status = "in_progress"

            # Get relevant source code
            source_code = self._get_step_code(state, step)

            prompt = MIGRATION_PROMPT.format(
                source=state.source_framework,
                target=state.target_framework,
                code=source_code,
                context=self._get_context(state)
            )

            response = self.llm.chat([
                {"role": "user", "content": prompt}
            ])

            # Extract all code blocks from response
            code_blocks = self._extract_all_code_blocks(response)

            if len(code_blocks) >= len(step.input_files) and len(step.input_files) > 1:
                # Multiple code blocks — try to match to input files
                matched = {}
                unmatched_blocks = list(code_blocks)

                # First pass: match by detected filename
                for block in code_blocks:
                    if block["filename"]:
                        for orig_filename in step.input_files:
                            new_filename = self._transform_filename(orig_filename, state.target_framework)
                            # Check if the detected filename matches (partial match OK)
                            if (block["filename"].lower().endswith(new_filename.lower().split("/")[-1]) or
                                new_filename.lower().endswith(block["filename"].lower().split("/")[-1])):
                                matched[new_filename] = block["code"]
                                if block in unmatched_blocks:
                                    unmatched_blocks.remove(block)
                                break

                # Second pass: positional matching for remaining files
                unmatched_files = [self._transform_filename(f, state.target_framework)
                                   for f in step.input_files if self._transform_filename(f, state.target_framework) not in matched]
                for i, new_filename in enumerate(unmatched_files):
                    if i < len(unmatched_blocks):
                        matched[new_filename] = unmatched_blocks[i]["code"]

                for new_filename, code in matched.items():
                    state.migrated_files[new_filename] = code
            else:
                # Single code block or single file — use first block
                migrated_code = code_blocks[0]["code"] if code_blocks else self._extract_code(response)
                for filename in step.input_files:
                    new_filename = self._transform_filename(filename, state.target_framework)
                    state.migrated_files[new_filename] = migrated_code

            step.status = "completed"
            step.result = response
            state.current_step += 1

        state.phase = Phase.VERIFICATION
        return state

    def _verify(self, state: MigrationState) -> MigrationState:
        """Phase 4: Verify migration results."""
        verification = {
            "files_migrated": len(state.migrated_files),
            "steps_completed": len([s for s in state.plan if s.status == "completed"]),
            "issues": [],
            "validations": []
        }

        # Verify each migrated file
        for filename, code in state.migrated_files.items():
            language = self._detect_language(filename)

            prompt = VERIFICATION_PROMPT.format(
                target=state.target_framework,
                language=language,
                code=code
            )

            response = self.llm.chat([
                {"role": "user", "content": prompt}
            ])

            try:
                result = self._parse_json(response)
                verification["validations"].append({
                    "file": filename,
                    "valid": result.get("valid", False),
                    "issues": result.get("issues", [])
                })
                if not result.get("valid", True):
                    verification["issues"].extend(result.get("issues", []))
            except Exception:
                verification["validations"].append({
                    "file": filename,
                    "valid": True,
                    "issues": []
                })

        state.verification_result = verification
        state.phase = Phase.COMPLETE
        return state

    def _detect_language(self, filename: str) -> str:
        """Detect language from filename."""
        ext_map = {
            ".py": "python",
            ".js": "javascript",
            ".ts": "typescript",
            ".java": "java",
            ".cs": "csharp",
            ".go": "go",
            ".rs": "rust"
        }
        for ext, lang in ext_map.items():
            if filename.endswith(ext):
                return lang
        return "unknown"

    def _parse_json(self, response: str) -> Dict:
        """Parse JSON from LLM response."""
        if "```json" in response:
            response = response.split("```json")[1].split("```")[0]
        elif "```" in response:
            response = response.split("```")[1].split("```")[0]
        return json.loads(response.strip())

    def _extract_code(self, response: str) -> str:
        """Extract first code block from response."""
        if "```" in response:
            parts = response.split("```")
            if len(parts) >= 2:
                code = parts[1]
                # Remove language identifier
                if "\n" in code:
                    first_line = code.split("\n")[0].strip()
                    if first_line in ("python", "javascript", "typescript", "java", "csharp", "go"):
                        code = code.split("\n", 1)[1] if "\n" in code else ""
                return code.strip()
        return response

    def _extract_all_code_blocks(self, response: str) -> List[Dict[str, str]]:
        """Extract all code blocks from response, with optional filename detection."""
        import re
        blocks = []
        # Split on code fences
        parts = response.split("```")
        # Code blocks are at odd indices (1, 3, 5, ...)
        for i in range(1, len(parts), 2):
            code = parts[i]
            # Check preceding text for filename hints
            preceding_text = parts[i - 1] if i > 0 else ""
            detected_filename = None

            # Look for filename patterns like "### Controllers/UserController.cs" or "**User.cs**" or "// filename.cs"
            filename_patterns = [
                r'[`*#/\\]+\s*([\w/\\]+\.(?:cs|java|py|js|ts|go|rs))\s*[`*]*',
                r'([\w/\\]+\.(?:cs|java|py|js|ts|go|rs))\s*$',
            ]
            for pattern in filename_patterns:
                match = re.search(pattern, preceding_text, re.MULTILINE | re.IGNORECASE)
                if match:
                    detected_filename = match.group(1).strip()
                    break

            # Remove language identifier from code
            if "\n" in code:
                first_line = code.split("\n")[0].strip()
                if first_line in ("python", "javascript", "typescript", "java", "csharp", "go"):
                    code = code.split("\n", 1)[1] if "\n" in code else ""

            # Also check the first comment line inside code for filename
            if not detected_filename:
                first_code_line = code.strip().split("\n")[0].strip() if code.strip() else ""
                comment_match = re.match(r'^(?://|#)\s*([\w/\\]+\.(?:cs|java|py|js|ts|go|rs))', first_code_line)
                if comment_match:
                    detected_filename = comment_match.group(1).strip()

            blocks.append({
                "filename": detected_filename,
                "code": code.strip()
            })
        return blocks

    def _get_step_code(self, state: MigrationState, step: MigrationStep) -> str:
        """Get source code for a migration step."""
        code_parts = []
        for filename in step.input_files:
            if filename in state.source_files:
                code_parts.append(f"# {filename}\n{state.source_files[filename]}")

        # If no specific files, include all
        if not code_parts:
            for filename, code in state.source_files.items():
                code_parts.append(f"# {filename}\n{code}")

        return "\n\n".join(code_parts)

    def _get_context(self, state: MigrationState) -> str:
        """Get context from previous steps."""
        completed = [s for s in state.plan if s.status == "completed"]
        if not completed:
            return "No previous steps completed."
        return "\n".join([f"Step {s.id}: {s.description}" for s in completed[-3:]])

    def _transform_filename(self, filename: str, target: str) -> str:
        """Transform filename for target framework."""
        transformations = {
            "fastapi": lambda f: f.replace(".js", ".py").replace("routes/", "routers/"),
            "express": lambda f: f.replace(".py", ".js").replace("routers/", "routes/"),
            "flask": lambda f: f.replace(".js", ".py"),
            "django": lambda f: f.replace(".js", ".py"),
            "dotnet-webapi": lambda f: f.replace(".java", ".cs").replace("controllers/", "Controllers/").replace("services/", "Services/").replace("models/", "Models/").replace("dto/", "Dto/"),
            "spring-boot": lambda f: f.replace(".cs", ".java").replace("Controllers/", "controllers/").replace("Services/", "services/").replace("Models/", "models/"),
        }
        transform = transformations.get(target.lower(), lambda f: f)
        return transform(filename)

File: labs/lab03-migration-workflow/typescript/src/agent.ts
/**
 * Migration Workflow Agent - Core Agent Implementation
 */

import type { LLMClient } from './llm-client.js';
import type {
  MigrationState,
  MigrationStep,
  AnalysisResult,
  VerificationResult,
} from './types.js';
import {
  createStep,
  setAnalysis,
  setPlan,
  updateStepStatus,
  addMigratedFile,
  incrementStep,
  setVerificationResult,
  addError,
} from './state.js';
import {
  ANALYSIS_PROMPT,
  PLANNING_PROMPT,
  MIGRATION_PROMPT,
  VERIFICATION_PROMPT,
  formatPrompt,
} from './prompts.js';

/**
 * Migration Agent that performs multi-step code migration
 */
export class MigrationAgent {
  private llm: LLMClient;

  constructor(llmClient: LLMClient) {
    this.llm = llmClient;
  }

  /**
   * Run the migration agent through all phases
   */
  async run(state: MigrationState): Promise<MigrationState> {
    while (state.phase !== 'complete') {
      state = await this.step(state);
      if (state.errors.length > 0) {
        break;
      }
    }
    return state;
  }

  /**
   * Execute one phase of the migration
   */
  private async step(state: MigrationState): Promise<MigrationState> {
    switch (state.phase) {
      case 'analysis':
        return this.analyze(state);
      case 'planning':
        return this.plan(state);
      case 'execution':
        return this.execute(state);
      case 'verification':
        return this.verify(state);
      default:
        return state;
    }
  }

  /**
   * Phase 1: Analyze source code
   */
  private async analyze(state: MigrationState): Promise<MigrationState> {
    const allAnalysis: Record<string, AnalysisResult> = {};

    for (const [filename, code] of Object.entries(state.sourceFiles)) {
      const prompt = formatPrompt(ANALYSIS_PROMPT, {
        source: state.sourceFramework,
        target: state.targetFramework,
        language: this.detectLanguage(filename),
        code,
      });

      const response = await this.llm.chat([
        { role: 'user', content: prompt },
      ]);

      try {
        allAnalysis[filename] = this.parseJson(response);
      } catch (e) {
        return addError(state, `Analysis failed for ${filename}: ${e}`);
      }
    }

    return setAnalysis(state, allAnalysis);
  }

  /**
   * Phase 2: Create migration plan
   */
  private async plan(state: MigrationState): Promise<MigrationState> {
    const prompt = formatPrompt(PLANNING_PROMPT, {
      analysis: JSON.stringify(state.analysis, null, 2),
      source: state.sourceFramework,
      target: state.targetFramework,
    });

    const response = await this.llm.chat([{ role: 'user', content: prompt }]);

    try {
      const planData = this.parseJson(response);
      const plan: MigrationStep[] = (planData.steps || []).map(
        (step: { id: number; description: string; input_files?: string[] }) =>
          createStep(step.id, step.description, step.input_files || [])
      );
      return setPlan(state, plan);
    } catch (e) {
      return addError(state, `Planning failed: ${e}`);
    }
  }

  /**
   * Phase 3: Execute migration steps
   */
  private async execute(state: MigrationState): Promise<MigrationState> {
    while (state.currentStep < state.plan.length) {
      const step = state.plan[state.currentStep];
      state = updateStepStatus(state, step.id, 'in_progress');

      const sourceCode = this.getStepCode(state, step);

      const prompt = formatPrompt(MIGRATION_PROMPT, {
        source: state.sourceFramework,
        target: state.targetFramework,
        code: sourceCode,
        context: this.getContext(state),
      });

      const response = await this.llm.chat([{ role: 'user', content: prompt }]);
      const codeBlocks = this.extractAllCodeBlocks(response);

      if (codeBlocks.length >= step.inputFiles.length && step.inputFiles.length > 1) {
        // Multiple code blocks — match to input files
        const matched: Record<string, string> = {};
        const unmatchedBlocks = [...codeBlocks];

        // First pass: match by detected filename
        for (const block of codeBlocks) {
          if (block.filename) {
            for (const origFilename of step.inputFiles) {
              const newFilename = this.transformFilename(origFilename, state.targetFramework);
              const blockBase = block.filename.toLowerCase().split('/').pop() || '';
              const targetBase = newFilename.toLowerCase().split('/').pop() || '';
              if (blockBase === targetBase || targetBase.endsWith(blockBase) || blockBase.endsWith(targetBase)) {
                matched[newFilename] = block.code;
                const idx = unmatchedBlocks.indexOf(block);
                if (idx >= 0) unmatchedBlocks.splice(idx, 1);
                break;
              }
            }
          }
        }

        // Second pass: positional matching for remaining
        const unmatchedFiles = step.inputFiles
          .map((f) => this.transformFilename(f, state.targetFramework))
          .filter((f) => !(f in matched));
        for (let i = 0; i < unmatchedFiles.length && i < unmatchedBlocks.length; i++) {
          matched[unmatchedFiles[i]] = unmatchedBlocks[i].code;
        }

        for (const [newFilename, code] of Object.entries(matched)) {
          state = addMigratedFile(state, newFilename, code);
        }
      } else {
        // Single block or single file
        const migratedCode = codeBlocks.length > 0 ? codeBlocks[0].code : this.extractCode(response);
        for (const filename of step.inputFiles) {
          const newFilename = this.transformFilename(filename, state.targetFramework);
          state = addMigratedFile(state, newFilename, migratedCode);
        }

        if (step.inputFiles.length === 0) {
          const defaultName = `migrated_step_${step.id}.${this.getExtension(state.targetFramework)}`;
          state = addMigratedFile(state, defaultName, migratedCode);
        }
      }

      state = updateStepStatus(state, step.id, 'completed', response);
      state = incrementStep(state);
    }

    return { ...state, phase: 'verification' };
  }

  /**
   * Phase 4: Verify migration results
   */
  private async verify(state: MigrationState): Promise<MigrationState> {
    const verification: VerificationResult = {
      filesMigrated: Object.keys(state.migratedFiles).length,
      stepsCompleted: state.plan.filter((s) => s.status === 'completed').length,
      issues: [],
      validations: [],
    };

    for (const [filename, code] of Object.entries(state.migratedFiles)) {
      const language = this.detectLanguage(filename);

      const prompt = formatPrompt(VERIFICATION_PROMPT, {
        target: state.targetFramework,
        language,
        code,
      });

      const response = await this.llm.chat([{ role: 'user', content: prompt }]);

      try {
        const result = this.parseJson(response);
        verification.validations.push({
          file: filename,
          valid: result.valid ?? true,
          issues: result.issues ?? [],
        });
        if (!result.valid) {
          verification.issues.push(...(result.issues ?? []));
        }
      } catch {
        verification.validations.push({
          file: filename,
          valid: true,
          issues: [],
        });
      }
    }

    return setVerificationResult(state, verification);
  }

  /**
   * Detect language from filename
   */
  private detectLanguage(filename: string): string {
    const extMap: Record<string, string> = {
      '.py': 'python',
      '.js': 'javascript',
      '.ts': 'typescript',
      '.java': 'java',
      '.cs': 'csharp',
      '.go': 'go',
      '.rs': 'rust',
    };
    for (const [ext, lang] of Object.entries(extMap)) {
      if (filename.endsWith(ext)) {
        return lang;
      }
    }
    return 'unknown';
  }

  /**
   * Get file extension for framework
   */
  private getExtension(framework: string): string {
    const extMap: Record<string, string> = {
      fastapi: 'py',
      flask: 'py',
      django: 'py',
      express: 'js',
      nestjs: 'ts',
      hono: 'ts',
      'dotnet-webapi': 'cs',
      'spring-boot': 'java',
    };
    return extMap[framework.toLowerCase()] || 'txt';
  }

  /**
   * Parse JSON from LLM response
   */
  private parseJson(response: string): Record<string, unknown> {
    let jsonStr = response;
    if (jsonStr.includes('```json')) {
      jsonStr = jsonStr.split('```json')[1].split('```')[0];
    } else if (jsonStr.includes('```')) {
      jsonStr = jsonStr.split('```')[1].split('```')[0];
    }
    return JSON.parse(jsonStr.trim());
  }

  /**
   * Extract code block from response
   */
  private extractCode(response: string): string {
    if (response.includes('```')) {
      const parts = response.split('```');
      if (parts.length >= 2) {
        let code = parts[1];
        const firstLine = code.split('\n')[0].trim();
        if (
          ['python', 'javascript', 'typescript', 'java', 'csharp', 'go'].includes(
            firstLine
          )
        ) {
          code = code.includes('\n') ? code.split('\n').slice(1).join('\n') : '';
        }
        return code.trim();
      }
    }
    return response;
  }

  private extractAllCodeBlocks(response: string): Array<{ filename: string | null; code: string }> {
    const blocks: Array<{ filename: string | null; code: string }> = [];
    const parts = response.split('```');
    const fileExtPattern = /[`*#/\\]+\s*([\w/\\]+\.(?:cs|java|py|js|ts|go|rs))\s*[`*]*/gi;
    const fallbackPattern = /([\w/\\]+\.(?:cs|java|py|js|ts|go|rs))\s*$/gim;
    const commentPattern = /^(?:\/\/|#)\s*([\w/\\]+\.(?:cs|java|py|js|ts|go|rs))/i;

    for (let i = 1; i < parts.length; i += 2) {
      let code = parts[i];
      const precedingText = parts[i - 1] || '';
      let detectedFilename: string | null = null;

      // Look for filename in preceding text
      const match1 = fileExtPattern.exec(precedingText);
      fileExtPattern.lastIndex = 0;
      if (match1) {
        detectedFilename = match1[1].trim();
      } else {
        const match2 = fallbackPattern.exec(precedingText);
        fallbackPattern.lastIndex = 0;
        if (match2) {
          detectedFilename = match2[1].trim();
        }
      }

      // Remove language identifier
      const firstLine = code.split('\n')[0].trim();
      if (['python', 'javascript', 'typescript', 'java', 'csharp', 'go'].includes(firstLine)) {
        code = code.includes('\n') ? code.split('\n').slice(1).join('\n') : '';
      }

      // Check first code line for filename comment
      if (!detectedFilename) {
        const firstCodeLine = code.trim().split('\n')[0]?.trim() || '';
        const commentMatch = commentPattern.exec(firstCodeLine);
        if (commentMatch) {
          detectedFilename = commentMatch[1].trim();
        }
      }

      blocks.push({ filename: detectedFilename, code: code.trim() });
    }
    return blocks;
  }

  /**
   * Get source code for a migration step
   */
  private getStepCode(state: MigrationState, step: MigrationStep): string {
    const codeParts: string[] = [];
    for (const filename of step.inputFiles) {
      if (state.sourceFiles[filename]) {
        codeParts.push(`// ${filename}\n${state.sourceFiles[filename]}`);
      }
    }
    if (codeParts.length === 0) {
      for (const [filename, code] of Object.entries(state.sourceFiles)) {
        codeParts.push(`// ${filename}\n${code}`);
      }
    }
    return codeParts.join('\n\n');
  }

  /**
   * Get context from previous steps
   */
  private getContext(state: MigrationState): string {
    const completed = state.plan.filter((s) => s.status === 'completed');
    if (completed.length === 0) {
      return 'No previous steps completed.';
    }
    return completed
      .slice(-3)
      .map((s) => `Step ${s.id}: ${s.description}`)
      .join('\n');
  }

  /**
   * Transform filename for target framework
   */
  private transformFilename(filename: string, target: string): string {
    const transformations: Record<string, (f: string) => string> = {
      fastapi: (f) => f.replace('.js', '.py').replace('routes/', 'routers/'),
      express: (f) => f.replace('.py', '.js').replace('routers/', 'routes/'),
      flask: (f) => f.replace('.js', '.py'),
      django: (f) => f.replace('.js', '.py'),
      hono: (f) => f.replace('.py', '.ts').replace('routers/', 'routes/'),
      nestjs: (f) => f.replace('.py', '.ts').replace('routers/', 'controllers/'),
      'dotnet-webapi': (f) => f.replace('.java', '.cs').replace('controllers/', 'Controllers/').replace('services/', 'Services/').replace('models/', 'Models/').replace('dto/', 'Dto/'),
      'spring-boot': (f) => f.replace('.cs', '.java').replace('Controllers/', 'controllers/').replace('Services/', 'services/').replace('Models/', 'models/'),
    };
    const transform = transformations[target.toLowerCase()] || ((f) => f);
    return transform(filename);
  }
}
